### PR TITLE
Overmap exoplanet generation, ported from Bay.

### PR DESCRIPTION
**Pull request by NonQueueingMatt**
To-do:

- [x] figure out why planets don't generate fully for whatever reason
- [ ] replace gas vars on tiles with lists
- [x] go through all the todos I left around
- [ ] figure out the graphite stuff
- [ ] add the trees maybe

### CODE REVIEW

**Comment by NonQueueingMatt** [comment]
!review

**Review by Alberyk — APPROVED** [review]

File: code/modules/overmap/exoplanets/random_map.dm [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+		var/image/grass = overlay_image('icons/obj/flora/greygrass.dmi', "grass_[grass_num]", color, RESET_COLOR)
+		grass.underlays += overlay_image('icons/obj/flora/greygrass.dmi', "grass_[grass_num]_shadow", null, RESET_COLOR)
+		LAZYSET(grass_cache, grass_num, grass)
+	return grass_cache[grass_num]
+
+/datum/random_map/noise/exoplanet/proc/spawn_flora(var/turf/T, var/big)
+	if(big)
+		if(LAZYLEN(big_flora_types))
+			new /obj/machinery/portable_atmospherics/hydroponics/soil/invisible(T, pick(big_flora_types), 1)
+		for(var/turf/neighbor in RANGE_TURFS(1,T))
```
Comment (Geevies) [inline code/modules/overmap/exoplanets/random_map.dm:114]:
```suggestion
		for(var/turf/neighbor as anything in RANGE_TURFS(1,T))
```

File: code/modules/overmap/exoplanets/turfs.dm [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+			to_chat(user,"<span class='notice'>You dig a deep pit.</span>")
+			new /obj/structure/pit(src)
+			diggable = 0
+		else
+			to_chat(user,"<span class='notice'>You stop shoveling.</span>")
+	else if(istype(C, /obj/item/stack/tile))
+		var/obj/item/stack/tile/T = C
+		if(T.use(1))
+			playsound(src, 'sound/items/Deconstruct.ogg', 80, 1)
+			ChangeTurf(/turf/simulated/floor, FALSE, FALSE, TRUE)
```
Comment (Geevies) [inline code/modules/overmap/exoplanets/turfs.dm:37]: shouldn't this adopt the turf type of the stack

File: code/modules/overmap/exoplanets/turfs.dm [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+		if(2)
+			if(prob(40))
+				ChangeTurf(get_base_turf_by_area(src))
+
+/turf/simulated/floor/exoplanet/Initialize()
+	. = ..()
+	update_icon(1)
+
+/turf/simulated/floor/exoplanet/update_icon(var/update_neighbors)
+	overlays.Cut()
```
Comment (Geevies) [inline code/modules/overmap/exoplanets/turfs.dm:54]:
```suggestion
	cut_overlays()
```

File: code/modules/overmap/exoplanets/turfs.dm [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+	if(LAZYLEN(decals))
+		overlays += decals
```
Comment (Geevies) [inline code/modules/overmap/exoplanets/turfs.dm:56]: add_overlay or add_overlays can't remember exactly

File: code/modules/overmap/exoplanets/turfs.dm [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+	O.desc = "You need to come over there to take a better look."
+
+/turf/simulated/planet_edge/CollidedWith(atom/movable/A)
+	. = ..()
+	var/obj/effect/overmap/visitable/sector/exoplanet/E = map_sectors["[z]"]
+	if(!istype(E))
+		return
+	if(E.planetary_area && istype(loc, world.area))
+		ChangeArea(src, E.planetary_area)
+		//TODO: Check if that should be done at the forcemove instead
```
Comment (Geevies) [inline code/modules/overmap/exoplanets/turfs.dm:210]: remember this TODO

File: code/modules/overmap/ships/engines/gas_thruster.dm [hunk 6]
```diff
@@ -97,13 +97,14 @@
...
-	for(var/ship in SSshuttle.ships)
-		var/obj/effect/overmap/visitable/ship/S = ship
-		if(S.check_ownership(src))
-			S.engines |= controller
-			if(dir != S.fore_dir)
-				stat |= BROKEN
-			break
+	if(length(SSshuttle.shuttle_areas) && !length(SSshuttle.shuttles_to_initialize) && SSshuttle.init_state == SS_INITSTATE_DONE)
+		for(var/ship in SSshuttle.ships)
+			var/obj/effect/overmap/visitable/ship/S = ship
```
Comment (Geevies) [inline code/modules/overmap/ships/engines/gas_thruster.dm:102]:
```suggestion
		for(var/obj/effect/overmap/visitable/ship/S as anything in SSshuttle.ships)
```

File: code/names.dm [hunk 7]
```diff
@@ -10,5 +10,20 @@ var/list/last_names = file2list("config/names/last.txt")
...
\ No newline at end of file
+//would include in rsc if ' was used
+
+var/list/greek_letters = list("Alpha", "Beta", "Gamma", "Delta",
+	"Epsilon", "Zeta", "Eta", "Theta", "Iota", "Kappa", "Lambda", "Mu",
+	"Nu", "Xi", "Omicron", "Pi", "Rho", "Sigma", "Tau", "Upsilon", "Phi",
+	"Chi", "Psi", "Omega")
+
+/proc/generate_system_name()
+	return "[pick("Gilese","GSC", "Luyten", "GJ", "HD", "SCGECO")][prob(10) ? " Eridani" : ""] [rand(100,999)]"
```
Comment (Geevies) [inline code/names.dm:23]: these are aurora friendly names right

**Review by Geevies — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/simple_animal/hostile/slug.dm [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+/mob/living/simple_animal/hostile/slug/Life()
+	. = ..()
+	if(. && istype(src.loc, /obj/item/holder) && isliving(src.loc.loc)) //We in somebody
+		var/mob/living/L = src.loc.loc
+		if(src.loc in L.get_visible_implants(0))
+			if(prob(1))
+				to_chat(L, "<span class='warning'>You feel strange as \the [src] pulses...</span>")
+			var/datum/reagents/R = L.reagents
+			R.add_reagent(/decl/reagent/cryptobiolin, 0.5)
```
[the remark this answers is not in the document]
Comment (NonQueueingMatt) [reply code/modules/mob/living/simple_animal/hostile/slug.dm:73]: i am just removing this fucking thing

File: code/modules/overmap/exoplanets/exoplanet.dm [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+
+	world.maxz++
+	forceMove(locate(1,1,world.maxz))
+
+	if(LAZYLEN(possible_themes))
+		var/datum/exoplanet_theme/T = pick(possible_themes)
+		themes += new T
+
+	for(var/T in subtypesof(/datum/map_template/ruin/exoplanet))
+		var/datum/map_template/ruin/exoplanet/ruin = T
```
[the remark this answers is not in the document]
Comment (NonQueueingMatt) [reply code/modules/overmap/exoplanets/exoplanet.dm:80]: I don't follow, it's not an object until it's newed

File: code/modules/overmap/exoplanets/turfs.dm (hunk 2 above)
Comment (NonQueueingMatt) [reply code/modules/overmap/exoplanets/turfs.dm:37]: Don't see a variable that lets me do this

**Review by Geevies — APPROVED** [review]

**Pull request merged**